finetuning-llama-7b-on-l4/README.md

```diff
@@ -0,0 +1,381 @@
+```
+
+Allow the K8s Service Account `l4-demo` in the `default` namespace to use the Google Service Account:
+```bash
+gcloud iam service-accounts add-iam-policy-binding ${SERVICE_ACCOUNT} \
+  --role roles/iam.workloadIdentityUser \
+  --member "serviceAccount:${PROJECT_ID}.svc.id.goog[default/l4-demo]"
+```
+
+Create a new K8s Service Account:
+```bash
+kubectl apply -f - << EOF
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: l4-demo
+  namespace: default
+  annotations:
+    iam.gke.io/gcp-service-account: l4-demo@$PROJECT_ID.iam.gserviceaccount.com
+EOF
+```
+
+Hugging face requires authentication to download the[ Llama 2 7B HF](https://huggingface.co/meta-llama/Llama-2-7b-hf) model, which means an access token is required to download the model.
+
+You can get your access token from [huggingface.com > Settings > Access Tokens](https://huggingface.co/settings/tokens). Afterwards, set your HuggingFace token as an environment variable:
+```bash
+export HF_TOKEN=<paste-your-own-token>
+```
+
+Create a Secret to store your HuggingFace token which will be used by the K8s job:
+```bash
+kubectl create secret generic l4-demo --from-literal="HF_TOKEN=$HF_TOKEN"
+```
+
+Run a K8s Job to download the the Llama 2 7B model to the bucket which is mounted to the K8s Pod:
+```bash
+kubectl apply -f - << EOF
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: model-loader
+  namespace: default
+spec:
+  template:
+    metadata:
+      annotations:
+        kubectl.kubernetes.io/default-container: loader
+        gke-gcsfuse/volumes: "true"
+        gke-gcsfuse/memory-limit: 400Mi
+        gke-gcsfuse/ephemeral-storage-limit: 30Gi
+    spec:
+      restartPolicy: OnFailure
+      containers:
+      - name: loader
+        image: python:3.11
+        command:
+        - /bin/bash
+        - -c
+        - |
+          cat > download-model.py << EOF
+          from huggingface_hub import snapshot_download
+          model_id="meta-llama/Llama-2-7b-hf"
+          snapshot_download(repo_id=model_id, local_dir="/gcs-mount/llama2-7b",
+                            local_dir_use_symlinks=False, revision="main",
+                            ignore_patterns=["*.safetensors", "model.safetensors.index.json"])
+          EOF
+          pip install huggingface_hub
+          python3 download-model.py
+        imagePullPolicy: IfNotPresent
+        env:
+        # kubectl create secret generic l4-demo --from-literal="HF_TOKEN=$HF_TOKEN"
+        - name: HUGGING_FACE_HUB_TOKEN
+          valueFrom:
+            secretKeyRef:
+              name: l4-demo
+              key: HF_TOKEN
+        volumeMounts:
+        - name: gcs-fuse-csi-ephemeral
+          mountPath: /gcs-mount
+      serviceAccountName: l4-demo
+      volumes:
+      - name: gcs-fuse-csi-ephemeral
+        csi:
+          driver: gcsfuse.csi.storage.gke.io
+          volumeAttributes:
+            bucketName: $BUCKET_NAME
+            mountOptions: "implicit-dirs"
```

Comment:
oh I see what you're saying, there are 2 jobs so could re-use the same PV that uses GCS fuse across the 2 jobs. I think I rather skip doing that for now since I had a lot of trouble getting GCS fuse to work so don't want to mess around with it too much.

Reply:
I think this will help because then I no longer need a variable in the job spec itself. I'm going to try this. Great suggestion.